### PR TITLE
Improve docker container reliability

### PR DESCRIPTION
**Pull request by etpinard**
@perigee @mdtusz @alexcjohnson @bpostlethwaite @cldougl 

This PR does not fix https://github.com/plotly/plotly.js/issues/62 and some local issues (ask @mdtusz :smiley_cat: ), but removes a significant amount of test-to-test randomness in the image test suite --- especially on CircleCI.

### CODE REVIEW

File: circle.yml [hunk 1]
```diff
@@ -12,11 +12,11 @@ machine:
...
 dependencies:
   pre:
-    - docker pull plotly/imageserver:latest
+    - docker pull plotly/imageserver:candidate
   post:
     - npm run cibuild
-    - docker run -d --name myimageserver -v $PWD:/var/www/streambed/image_server/plotly.js -p 9010:9010 plotly/imageserver:latest; sleep 20
-
+    - docker run -d --name myimageserver -v $PWD:/var/www/streambed/image_server/plotly.js -p 9010:9010 plotly/imageserver:candidate
+    - wget --server-response --spider --tries=8 --retry-connrefused http://localhost:9010/ping
```
Comment (etpinard) [inline circle.yml]: wait until image server returns a `pong` before starting the tests.

File: circle.yml [hunk 2]
```diff
@@ -12,11 +12,11 @@ machine:
 
 dependencies:
   pre:
-    - docker pull plotly/imageserver:latest
+    - docker pull plotly/imageserver:candidate
```
Comment (etpinard) [inline circle.yml]: @perigee 

TODO:
- [ ] update latest container on dockerhub

File: tasks/baseline.sh [hunk 3]
```diff
@@ -12,8 +12,8 @@ CONTAINER_NAME="imagetest"  # same as in docker-compose.yml
 CMD=(
     "cd /var/www/streambed/image_server/plotly.js &&"
     "cp -f test/image/index.html ../server_app/index.html &&"
-    "monit restart nw1 &&"
-    "sleep 5 &&"
+    "supervisorctl restart nw1 &&"
```
Comment (etpinard) [inline tasks/baseline.sh]: bye bye monit.

File: test/image/compare_pixels_test.js [hunk 4]
```diff
@@ -35,15 +40,34 @@ function runAll () {
...
+            testMock(mocks[cnt++], t);
+        }
+
+        t.plan(mocks.length);
 
-        for (var i = 0; i < files.length; i ++) {
-            testMock(files[i], t);
+        for(var i = 0; i < mocks.length; i++) {
+            setTimeout(testFunction,
+                BASE_TIMEOUT * Math.floor(i / BATCH_SIZE) * BATCH_SIZE);
```
Comment (etpinard) [inline test/image/compare_pixels_test.js]: Send 5 requests at a time, not 235.

File: test/image/compare_pixels_test.js [hunk 5]
```diff
@@ -35,15 +40,34 @@ function runAll () {
...
+         * - gl3d_bunny-hull (1)
+         * - polar_scatter (1)
          */
-        t.plan(files.length - 40);
+        var mocks = allMocks.filter(function(mock) {
+            return !(
+                mock === 'font-wishlist.json' ||
+                mock.indexOf('gl2d') !== -1 ||
+                mock === 'gl3d_bunny-hull.json' ||
+                mock === 'polar_scatter.json'
```
Comment (etpinard) [inline test/image/compare_pixels_test.js]: this mock is also showing some test-to-test randomness, let's skip it for now.

Comment (cldougl) [inline test/image/compare_pixels_test.js]: :+1:

**Comment by etpinard** [comment]
Robust CircleCI testing:

File: circle.yml (hunk 2 above)
Comment (perigee) [inline circle.yml]: pushed new latest, you can still use latest though ;)

**Comment by perigee** [comment]
:dancer:  for docker related part ;D

File: tasks/baseline.sh (hunk 3 above)
Comment (perigee) [inline tasks/baseline.sh]: :+1:

File: test/image/compare_pixels_test.js (hunk 4 above)
Comment (bpostlethwaite) [inline test/image/compare_pixels_test.js]: Man - I really assumed we were already batching them. Can't believe we went this long without failures! Good to know the image_server can work with such a huge onslaught :)

File: circle.yml [hunk 6]
```diff
@@ -15,8 +15,8 @@ dependencies:
     - docker pull plotly/imageserver:latest
   post:
     - npm run cibuild
-    - docker run -d --name myimageserver -v $PWD:/var/www/streambed/image_server/plotly.js -p 9010:9010 plotly/imageserver:latest; sleep 20
-
+    - docker run -d --name myimageserver -v $PWD:/var/www/streambed/image_server/plotly.js -p 9010:9010 plotly/imageserver:latest
+    - wget --server-response --spider --tries=8 --retry-connrefused http://localhost:9010/ping
```
Comment (etpinard) [inline circle.yml]: An even better script would restart `nw1` if the image server doesn't return a `pong` after 8 tries.

Looks like the latest container is reliable enough that we don't need this now. But, maybe later.

File: test/image/compare_pixels_test.js (hunk 5 above)
Comment (bpostlethwaite) [inline test/image/compare_pixels_test.js]: Can we add a corresponding issue for these and if so could you link to it in the code here with a 
"See issue [issuelink]"

**Comment by bpostlethwaite** [comment]
Let's issue up those missed mocks and link to them in the script. Once that is in place it looks :dancer: to me!

File: test/image/compare_pixels_test.js (hunk 5 above)
Comment (bpostlethwaite) [inline test/image/compare_pixels_test.js]: Haha Rodger dodger!

**Comment by bpostlethwaite** [comment]
Blind as a bat! Great :dancer:

**Pull request merged**